components/script/dom/htmlscriptelement.rs

```diff
@@ -148,17 +171,47 @@ impl AsyncResponseListener for ScriptContext {
     }
 
     fn response_complete(&mut self, status: Result<(), String>) {
-        let load = status.map(|_| {
-            let data = mem::replace(&mut self.data, vec!());
+        let script = status.map(|_| {
             let metadata = self.metadata.take().unwrap();
-            (metadata, data)
+
+            // TODO(#9185): implement encoding determination.
+            // Step 8.
+            let source_text = UTF_8.decode(&self.data, DecoderTrap::Replace).unwrap();
+            ClassicScript {
+                source_text: source_text.into(),
+                external: true,
+                url: metadata.final_url,
+            }
         });
         let elem = self.elem.root();
-        // TODO: maybe set this to None again after script execution to save memory.
-        *elem.load.borrow_mut() = Some(ScriptOrigin::External(load));
-        elem.ready_to_be_parser_executed.set(true);
 
+        match self.kind {
+            ExternalScriptKind::ParsingBlocking => {
+                elem.ready_to_be_parser_executed.set(true);
+            }
+            ExternalScriptKind::Deferred => {
+                let document = document_from_node(&*elem);
+                document.deferred_script_loaded(&elem, script);
+            }
+            ExternalScriptKind::AsapInOrder => {
+                // XXX associated with the node document of the script element
+                //     **at the time the prepare a script algorithm started**.
+                let document = document_from_node(&*elem);
+                document.asap_in_order_script_loaded(&elem, script);
+            },
+            ExternalScriptKind::Asap => {
+                // XXX associated with the node document of the script element
+                //     **at the time the prepare a script algorithm started**.
+                let document = document_from_node(&*elem);
+                document.process_asap_script(&elem, script);
+            },
+        }
+
+        // This might have caused one of the conditions for spinning the event
+        // loop in the-end to have changed.
         let document = document_from_node(elem.r());
+        document.continue_the_end();
+
         document.finish_load(LoadType::Script(self.url.clone()));
```

Comment:
Should we invoke finish_load before continue_the_end? The script counts as something blocking the document load event, so it has an effect on what that method does.